lib/datadog/appsec/contrib/rack/request_middleware.rb

```diff
@@ -49,6 +54,50 @@ def call(env)
             AppSec::Event.record(*both_response.map { |_action, event| event }) if both_response.any?
 
             request_return
+          ensure
+            add_waf_runtime_tags(context)
+            request!
+          end
+
+          private
+
+          def first_request?
+            @first_request
+          end
+
+          def request!
+            @first_request = false
+          end
+
+          def active_trace
+            return unless defined?(Datadog::Tracing) && Datadog::Tracing.respond_to?(:active_span)
+
+            Datadog::Tracing.active_trace
+          end
+
+          def add_appsec_tags
+            return unless active_trace
+
+            active_trace.set_tag('_dd.appsec.enabled', 1)
+            active_trace.set_tag('_dd.runtime_family', 'ruby')
+            active_trace.set_tag('_dd.appsec.waf.version', Datadog::AppSec::WAF::VERSION::BASE_STRING)
+
+            if @processor.ruleset_info
+              active_trace.set_tag('_dd.appsec.event_rules.version', @processor.ruleset_info[:version])
+              if first_request?
+                active_trace.set_tag('_dd.appsec.event_rules.loaded', @processor.ruleset_info[:loaded].to_f)
+                active_trace.set_tag('_dd.appsec.event_rules.error_count', @processor.ruleset_info[:failed].to_f)
+                active_trace.set_tag('_dd.appsec.event_rules.errors', JSON.dump(@processor.ruleset_info[:errors]))
+                active_trace.set_tag('_dd.appsec.event_rules.addresses', JSON.dump(@processor.addresses))
+                active_trace.keep!
```

Comment:
Minor: This is a bit weird since it can override an explicit customer `reject!`. Should this check first if `sampled?` is set before calling the `keep!`?

Reply:
I believe if this span represents a security alert, it would override sampling, even if the user explicit doesn't want it retained, is that correct @lloeki?

I can see this going in two ways, depending on what the Appsec team wants here: one might not know they a security issue in their endpoint and not informing the user because they chose to not perform tracing collection could be seem as incorrect. But at the same time, if the endpoint is clearly inconsequential (e.g. a health-check endpoint, or internal-only route) then I can see the argument to respect `keep!` and `reject!`.

If you want to check for explicit user sampling, I'd suggest adding a new method to `TraceOperation`:
```ruby
      # The user has explicitly set the tracing decision for this operation.
      def user_sampling?
        sampling_priority == Sampling::Ext::Priority::USER_KEEP || sampling_priority == Sampling::Ext::Priority::USER_REJECT
      end
```
I don't think there's a good way to know this today with the currently exposed API in `TraceOperation`, so a new method would work.

Reply:
Good question and astute analysis. I need to confirm with the team what the expectation is, although I think it is expected to be retained in spite of being rejected by the user.

Reply:
Answer: currently undefined. There are pros and cons, but it seems like the user having the final word weighs in big.

Reply:
Ah I was confused: this only happens on the first request, in order to send telemetry, irrespective of attacks. In this case it is mandatory to send it.

Reply:
@marcotc would it make sense for `TraceOperation` to emit a log warning when a trace has gone from `USER_REJECT` to `USER_KEEP`? At least in that case customers would have visibility on why such a weird thing happened.